misc/multistream-select/src/negotiated.rs

```diff
@@ -0,0 +1,338 @@
+
+use bytes::BytesMut;
+use crate::protocol::{Protocol, MessageReader, Message, Version, ProtocolError};
+use futures::{prelude::*, Async, try_ready};
+use log::debug;
+use tokio_io::{AsyncRead, AsyncWrite};
+use std::{mem, io, fmt, error::Error};
+
+/// An I/O stream that has settled on an (application-layer) protocol to use.
+///
+/// A `Negotiated` represents an I/O stream that has _settled_ on a protocol
+/// to use. In particular, it is not implied that all of the protocol negotiation
+/// frames have yet been sent and / or received, just that the selected protocol
+/// is fully determined. This is to allow the last protocol negotiation frames
+/// sent by a peer to be combined in a single write, possibly piggy-backing
+/// data from the negotiated protocol on top.
+///
+/// Reading from a `Negotiated` I/O stream that still has pending negotiation
+/// protocol data to send implicitly triggers flushing of all yet unsent data.
+pub struct Negotiated<TInner> {
+    state: State<TInner>
+}
+
+/// A `Future` that waits on the completion of protocol negotiation.
+pub struct NegotiatedComplete<TInner> {
+    inner: Option<Negotiated<TInner>>
+}
+
+impl<TInner: AsyncRead + AsyncWrite> Future for NegotiatedComplete<TInner> {
+    type Item = Negotiated<TInner>;
+    type Error = NegotiationError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut io = self.inner.take().expect("NegotiatedFuture called after completion.");
+        if io.poll()?.is_not_ready() {
+            self.inner = Some(io);
+            return Ok(Async::NotReady)
+        }
+        return Ok(Async::Ready(io))
+    }
+}
+
+impl<TInner> Negotiated<TInner> {
+    /// Creates a `Negotiated` in state [`State::Complete`], possibly
+    /// with `remaining` data to be sent.
+    pub(crate) fn completed(io: TInner, remaining: BytesMut) -> Self {
+        Negotiated { state: State::Completed { io, remaining } }
+    }
+
+    /// Creates a `Negotiated` in state [`State::Expecting`] that is still
+    /// expecting confirmation of the given `protocol`.
+    pub(crate) fn expecting(io: MessageReader<TInner>, protocol: Protocol) -> Self {
+        Negotiated { state: State::Expecting { io, protocol } }
+    }
+
+    /// Polls the `Negotiated` for completion.
+    fn poll(&mut self) -> Poll<(), NegotiationError>
+    where
+        TInner: AsyncRead + AsyncWrite
+    {
+        // Flush any pending negotiation data.
+        match self.poll_flush() {
+            Ok(Async::Ready(())) => {},
+            Ok(Async::NotReady) => return Ok(Async::NotReady),
+            Err(e) => {
+                // If the remote closed the stream, it is important to still
+                // continue reading the data that was sent, if any.
+                if e.kind() != io::ErrorKind::WriteZero {
+                    return Err(e.into())
+                }
+            }
+        }
+
+        if let State::Completed { remaining, .. } = &mut self.state {
```

Comment:
Nit-pick: this block could be moved to line 82 (`Ok(Async::Ready(())) => {}`) for clarity.

Reply:
Hm, I don't think so, because this should also happen if `poll_flush` returned an `Err` with `WriteZero`.